File: cpp/src/processing/tracking_processor.cpp
#include "headcannon/processing/tracking_processor.h"
#include "headcannon/math/deadzone_utils.h"
#include "headcannon/math/smoothing_utils.h"
#include "headcannon/math/angle_utils.h"

namespace headcannon {

TrackingPose TrackingProcessor::Process(float yaw, float pitch, float roll, bool is_remote_connection, float delta_time) {
    // Step 1: Apply center offset
    m_centerManager.ApplyOffset(yaw, pitch, roll);

    // Step 2: Apply deadzone
    yaw = static_cast<float>(math::ApplyDeadzone(yaw, m_deadzone.yaw));
    pitch = static_cast<float>(math::ApplyDeadzone(pitch, m_deadzone.pitch));
    roll = static_cast<float>(math::ApplyDeadzone(roll, m_deadzone.roll));

    // Step 3: Apply smoothing
    double effective_smoothing = math::GetEffectiveSmoothing(m_smoothingFactor, is_remote_connection);

    if (!m_hasSmoothedValue) {
        // First frame, snap to target
        m_smoothedYaw = yaw;
        m_smoothedPitch = pitch;
        m_smoothedRoll = roll;
        m_hasSmoothedValue = true;
    } else {
        m_smoothedYaw = math::Smooth(m_smoothedYaw, static_cast<double>(yaw), effective_smoothing, static_cast<double>(delta_time));
        m_smoothedPitch = math::Smooth(m_smoothedPitch, static_cast<double>(pitch), effective_smoothing, static_cast<double>(delta_time));
        m_smoothedRoll = math::Smooth(m_smoothedRoll, static_cast<double>(roll), effective_smoothing, static_cast<double>(delta_time));
    }

    // Step 4: Apply sensitivity
    float out_yaw = static_cast<float>(m_smoothedYaw) * m_sensitivity.yaw;
    float out_pitch = static_cast<float>(m_smoothedPitch) * m_sensitivity.pitch;
    float out_roll = static_cast<float>(m_smoothedRoll) * m_sensitivity.roll;

    if (m_sensitivity.invert_yaw) out_yaw = -out_yaw;
    if (m_sensitivity.invert_pitch) out_pitch = -out_pitch;
    if (m_sensitivity.invert_roll) out_roll = -out_roll;

    // Step 5: Apply limits
    out_yaw = math::Clamp(out_yaw, m_limits.yaw_min, m_limits.yaw_max);
    out_pitch = math::Clamp(out_pitch, m_limits.pitch_min, m_limits.pitch_max);
    out_roll = math::Clamp(out_roll, m_limits.roll_min, m_limits.roll_max);

    return TrackingPose(out_yaw, out_pitch, out_roll);
}

void TrackingProcessor::Recenter() {
    m_centerManager.SetCenter(
        static_cast<float>(m_smoothedYaw),
        static_cast<float>(m_smoothedPitch),
        static_cast<float>(m_smoothedRoll)
    );
}

void TrackingProcessor::RecenterTo(float yaw, float pitch, float roll) {
    m_centerManager.SetCenter(yaw, pitch, roll);
    m_smoothedYaw = 0.0;
    m_smoothedPitch = 0.0;
    m_smoothedRoll = 0.0;
}

void TrackingProcessor::Reset() {
    m_centerManager.Reset();
    m_smoothedYaw = 0.0;
    m_smoothedPitch = 0.0;
    m_smoothedRoll = 0.0;
    m_hasSmoothedValue = false;
}

}  // namespace headcannon

File: csharp/src/CameraUnlock.Core/Processing/TrackingProcessor.cs
using CameraUnlock.Core.Data;
using CameraUnlock.Core.Math;

namespace CameraUnlock.Core.Processing
{
    /// <summary>
    /// Complete tracking data processing pipeline.
    /// Pipeline: raw -> offset -> deadzone -> smooth -> sensitivity
    /// </summary>
    public sealed class TrackingProcessor : ITrackingProcessor
    {
        private readonly CenterOffsetManager _centerManager = new CenterOffsetManager();

        // Smoothed values (stored as doubles for precision)
        private double _smoothedYaw;
        private double _smoothedPitch;
        private double _smoothedRoll;
        private bool _hasSmoothedValue;

        /// <summary>
        /// Sensitivity settings.
        /// </summary>
        public SensitivitySettings Sensitivity { get; set; } = SensitivitySettings.Default;

        /// <summary>
        /// Deadzone settings.
        /// </summary>
        public DeadzoneSettings Deadzone { get; set; } = DeadzoneSettings.None;

        /// <summary>
        /// Smoothing factor (0 = instant, 1 = very slow).
        /// </summary>
        public float SmoothingFactor { get; set; } = 0f;

        /// <summary>
        /// The center offset manager for recentering.
        /// </summary>
        public CenterOffsetManager CenterManager => _centerManager;

        /// <summary>
        /// Gets the current smoothed rotation values.
        /// </summary>
        /// <param name="yaw">Output: smoothed yaw.</param>
        /// <param name="pitch">Output: smoothed pitch.</param>
        /// <param name="roll">Output: smoothed roll.</param>
        public void GetSmoothedRotation(out float yaw, out float pitch, out float roll)
        {
            yaw = (float)_smoothedYaw;
            pitch = (float)_smoothedPitch;
            roll = (float)_smoothedRoll;
        }

#if NETSTANDARD2_0
        /// <summary>
        /// Gets the current smoothed rotation values (tuple version for modern runtimes).
        /// </summary>
        public (float Yaw, float Pitch, float Roll) SmoothedRotation =>
            ((float)_smoothedYaw, (float)_smoothedPitch, (float)_smoothedRoll);
#endif

        /// <summary>
        /// Processes a raw tracking pose through the full pipeline.
        /// </summary>
        public TrackingPose Process(TrackingPose rawPose, bool isRemoteConnection, float deltaTime)
        {
            if (!rawPose.IsValid)
            {
                return rawPose;
            }

            // Step 1: Apply center offset
            _centerManager.ApplyOffset(rawPose.Yaw, rawPose.Pitch, rawPose.Roll, out float yaw, out float pitch, out float roll);

            // Step 2: Apply deadzone
            yaw = (float)DeadzoneUtils.Apply(yaw, Deadzone.Yaw);
            pitch = (float)DeadzoneUtils.Apply(pitch, Deadzone.Pitch);
            roll = (float)DeadzoneUtils.Apply(roll, Deadzone.Roll);

            // Step 3: Apply smoothing
            float effectiveSmoothing = SmoothingUtils.GetEffectiveSmoothing(SmoothingFactor, isRemoteConnection);

            if (!_hasSmoothedValue)
            {
                // First frame, snap to target
                _smoothedYaw = yaw;
                _smoothedPitch = pitch;
                _smoothedRoll = roll;
                _hasSmoothedValue = true;
            }
            else
            {
                // Calculate smoothing factor once for all axes (avoids 3x Math.Exp calls)
                float t = SmoothingUtils.CalculateSmoothingFactor(effectiveSmoothing, deltaTime);
                _smoothedYaw = _smoothedYaw + (yaw - _smoothedYaw) * t;
                _smoothedPitch = _smoothedPitch + (pitch - _smoothedPitch) * t;
                _smoothedRoll = _smoothedRoll + (roll - _smoothedRoll) * t;
            }

            // Step 4: Apply sensitivity
            return new TrackingPose((float)_smoothedYaw, (float)_smoothedPitch, (float)_smoothedRoll, rawPose.TimestampTicks)
                .ApplySensitivity(Sensitivity);
        }

        /// <summary>
        /// Sets the current smoothed pose as the center.
        /// </summary>
        public void Recenter()
        {
            _centerManager.SetCenter((float)_smoothedYaw, (float)_smoothedPitch, (float)_smoothedRoll);
        }

        /// <summary>
        /// Sets a specific pose as the center (for recentering without prior data).
        /// </summary>
        public void RecenterTo(TrackingPose pose)
        {
            _centerManager.SetCenter(pose);
            _smoothedYaw = 0;
            _smoothedPitch = 0;
            _smoothedRoll = 0;
        }

        /// <summary>
        /// Resets the processor state.
        /// </summary>
        public void Reset()
        {
            _centerManager.Reset();
            _smoothedYaw = 0;
            _smoothedPitch = 0;
            _smoothedRoll = 0;
            _hasSmoothedValue = false;
        }
    }
}
